TUI feature request: grow text field vertically instead of scrolling #101

trevarj opened the pull request:
Modified text_fields.rs to expand vertically when reaching the end of the line.

@osa1 let me know if there's anything I should fix or look out for (maybe I didn't test enough areas).

Side by side diff may be easier to see the changes.

osa1 commented:
Awesome, thanks for the patch!

Great work overall. Before looking at the code in detail I tried a few things. It mostly looks good. Some comments below.

- If the text field grows vertically and then I delete things, the screen layout is not completely restored. In the screenshot below I typed something long and then deleted, and there's on line gap between the text field and in message area now.

![Screenshot_2020-04-04_06-56-27](https://user-images.githubusercontent.com/448274/78418066-740a3800-7641-11ea-9a4a-cba1a84b03cd.png)

- This feature should be optional, enabled with a line in the config.yml file. I think something like `text_field_wrap: true` would work. If it's not available or set to `false` then the feature is not enabled.

- On small screens we should fall back to the current behavior, which is more compact (always takes 1 row). When to fall back exactly? A few random ideas (feel free to suggest more):
    - If MsgArea is smaller than a fixed amount (e.g. 5 lines)
    - If MsgArea is smaller than the input field

Btw, have you seen [the TUI tests][1]? It'd be good to add some test about these, especially for resizing and rendering on small screens.

Thanks again!

[1]: https://github.com/osa1/tiny/blob/master/libtiny_tui/src/tests/mod.rs

trevarj commented:
Thanks for the feedback! I will try to implement these suggestions when I get some more time.

trevarj commented:
@osa1 Hey, I'm making some good progress, but I'm finding it difficult to add a configuration parameter...It doesn't look like there's a nice way to configure a setting on the TextField without passing down the config all the way from main.rs...Any ideas?

Also, I'm pretty overwhelmed with the test framework...I never really worked in testing a TUI before 😩 

I wasn't able to reproduce the problem you're seeing and mentioned above in your screenshot. You can see from the video below that I don't have that issue on my terminal emulator...I may need to try some other terminals.
Here's what I have at the moment: 

https://streamable.com/lxpt34

osa1 commented:
Hey @trevarj,

> I'm finding it difficult to add a configuration parameter...It doesn't look like there's a nice way to configure a setting on the TextField without passing down the config all the way from main.rs...Any ideas?

TUI has its own config format and parser, see `libtiny_tui/src/config.rs`. You need to parse the config field there (just update the `Config` struct in that file with a new `bool` field) and use it in `reload_config` in `libtiny_tui/src/tui.rs`. It should be too hard. No need to make any changes in `main.rs`.

> Also, I'm pretty overwhelmed with the test framework...I never really worked in testing a TUI before

TUI test library is really simple, have you seen some of the tests in `libtiny_tui/src/tests/mod.rs`? For example, this test in that file

```
#[test]
fn init_screen() {
    let mut tui = TUI::new_test(20, 4);
    tui.draw();

    #[rustfmt::skip]
    let screen =
        "|Any mentions to you |
         |will be listed here.|
         |                    |
         |mentions            |";
    expect_screen(screen, &tui, 20, 4);
}
```

creates a TUI and then compares the screen with the textual representation (`screen`) of it.

To update the TUI after it's initialized you can either send key press events using the `handle_input_event` method or add a method for adding stuff to the text field, and then use that.

> I wasn't able to reproduce the problem you're seeing and mentioned above in your screenshot

Here's how to produce it:

- Make your screen small so that you won't have to enter long text.
- Move to a tab where the message area is full of messages. (there shouldn't be any empty lines in the incoming messages)
- Enter some long text so that the input field grows to multiple rows.
- Now delete the text. You should end up with some empty space above the input field, as shown in my screenshot above.

---

Let me know if you have any more questions!

trevarj commented:
Which terminal do you use? Below is me following the steps 😟 
https://imgur.com/a/BaxT92o

osa1 commented:
Here's a demo:

![tiny_demo](https://user-images.githubusercontent.com/448274/78473407-224cd500-7749-11ea-955e-5a6519161bd6.gif)

trevarj commented:
I installed on linux and ran some testing...looks ok for me. Maybe you didn't pull the most recent changes?

trevarj commented on libtiny_tui/src/text_field.rs:
```diff
@@ -9,7 +9,8 @@ use termbox_simple::Termbox;
 // TODO: Make these settings
-const SCROLLOFF: i32 = 5;
+const SCROLL_OFF: i32 = 5;
+const OVERFLOW_WIDTH: i32 = 30;
```
might want to make this larger. not sure yet.

osa1 commented on libtiny_tui/src/config.rs:
```diff
@@ -14,6 +14,8 @@ pub use termbox_simple::*;
 pub(crate) struct Config {
     #[serde(default)]
     pub(crate) colors: Colors,
+    #[serde(default = "text_field_wrap_default")]
```
I think default for `bool` is already `false` so you could simply use `#[serde(default)]` and remove `text_field_wrap_default` function.

osa1 commented on libtiny_tui/src/tests/mod.rs:
```diff
@@ -170,3 +171,150 @@ fn small_screen_2() {
+     |osa1: aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa |
+     |mentions chat.freenode.net                                  |";
+    //                                               cursor space^
+    expect_screen(screen, &tui, 60, 30);
+}
+
+///
+/// Tests scroll (text wrap on) small screen
+///
+#[test]
+fn test_scroll_text_field_wrap_true() {
+    let mut tui = TUI::new_test(30, 10);
+    tui.set_text_field_wrap_test(true);
+
+    let server = "chat.freenode.net";
+    tui.new_server_tab(server);
+    tui.set_nick(server, "osa1");
+    // switch to server tab
+    tui.next_tab();
+
+    for _ in 0..65 {
+        let event = term_input::Event::Key(Key::Char('a'));
+        tui.handle_input_event(event);
+    }
+
+    tui.draw();
+
+    let screen = "|                              |
+     |                              |
+     |                              |
+     |                              |
+     |                              |
+     |                              |
+     |                              |
+     |                              |
+     |osa1: aaaaaaaaaaaaaaaaaaaaaaa |
+     |mentions chat.freenode.net    |";
+    //                 cursor space^
+    expect_screen(screen, &tui, 30, 10);
+}
```
Could you please make the examples smaller? No need for this large screen to test this feature. I think 10x5 screen should be enough.

It'd also be good to have some text in the messaging area while testing this, to make sure the text is rendered properly. For example, if there's a line above the input field it should be visible after text field grows one more row. Similarly if you remove text so that text field takes 1 row again the messaging area should again be resized and the live above the input field should again be right above of the input field.

If you make the screen sizes smaller it'll also take less effort to implement these tests.

trevarj replied:
I will work on adding some more tests like one to check for the bug you mentioned a few days ago.
I think that I need the larger screen, at least to test the wrapping, because it defaults back to scrolling mode when it's too small...

osa1 replied:
> I think that I need the larger screen, at least to test the wrapping, because it defaults back to scrolling mode when it's too small...

Hmm, right. I think you could make `OVERFLOW_WIDTH` a parameter and override it in tests. Even if we want to keep `OVERFLOW_WIDTH` constant I think you could at least make height smaller.

osa1 commented on libtiny_tui/src/text_field.rs:
```diff
@@ -19,12 +20,18 @@ pub(crate) struct TextField {
+    /// Whether or not scroll mode is currently on
+    scrolling_on: bool,
+    /// Amount of scroll in the input field
+    scroll: i32,
```
So this value is only useful when `scrolling_on` is `true`, right? If that's the case than I think using `scroll: Option<i32>` and removing `scroll_on` might be better. If `scroll` is `None` then we know that scrolling is off. Otherwise the value gives us the scroll value.

trevarj replied:
I think I can do that. Thanks for the tip!

osa1 commented on libtiny_tui/src/text_field.rs:
```diff
@@ -51,19 +58,24 @@ enum Mode {
+            max_lines,
+            text_field_wrap,
+            scrolling_on: !text_field_wrap,
+            scroll: 0,
```
This could be simplified if you use the idea I mention above.

trevarj commented:
> There's currently one inconsistency in how we split long lines in messaging area vs. in text field. In messaging area we only split lines at word boundaries (at whitespaces). That works really well I think and renders nicely. I wonder if we could do the same for the text field.

I did notice then when I first started playing with the tui to make my change. It seemed a little difficult for me to extract out, especially with how the autocomplete stuff is right now...Maybe some day I can implement it though.

> Could you investigate this?
> 
> Other than this bug I can't really see any problems. Great work!

Finished up all other suggestions. Just need to debug this bug and I will commit to the PR again!

osa1 commented:
Sorry for the delay, I've been busy with work and life. I'll try to take another look today.

osa1 commented on libtiny_tui/src/text_field.rs:
```diff
@@ -453,6 +489,21 @@ impl TextField {
 
+    /// Calculate how many lines of text will be in the textfield
+    /// based on the width of the widget
+    pub fn calculate_lines(&self) -> i32 {
```
`pub(crate)`

osa1 commented on libtiny_tui/src/text_field.rs:
```diff
@@ -9,7 +9,8 @@ use termbox_simple::Termbox;
 // TODO: Make these settings
-const SCROLLOFF: i32 = 5;
+const SCROLL_OFF: i32 = 5;
+const OVERFLOW_WIDTH: i32 = 36;
```
I don't understand why this is necessary. @trevarj could you document this please before I merge this PR?

trevarj replied:
I will add some documentation comments and rename `OVERFLOW_WIDTH`

trevarj replied:
Honestly, I don't even know what `SCROLL_OFF` does! I'm playing around with changing its value and it doesn't seem like it does anything.

osa1 replied:
> I'm playing around with changing its value and it doesn't seem like it does anything.

Yeah I tried this as well. Also tried just deleting it, and nothing broke! Let's remove it.

osa1 replied:
Ahh, you mean `SCROLL_OFF`. That's actually useful, I'll document it.

trevarj replied:
Never mind, I see what it does, but I can't explain it in words!

osa1 replied:
It's inspired by vim's [`scrolloff`](http://vimdoc.sourceforge.net/htmldoc/options.html#'scrolloff') ;-)

trevarj replied:
About `OVERFLOW_WIDTH`, I renamed it to `SCROLL_FALLBACK_WIDTH`. It's the value that is used when you have `text_field_wrap=true` and then resize the terminal. When the text_field's width is < 36 it will fall back to scroll mode.

osa1 replied:
Why is this necessary? If the terminal width is 20, with enough height it should be possible to use wrapping. Is there a problem with this?

I don't know if you've changed the code, but in the version I've reviewed I removed this variable (and the conditional) and everything worked as before -- I was able to use wrapping on a small terminal. So I'm also not sure if it works as intended.

trevarj replied:
Oh, I just thought you wanted it to fall back to scrolling when the terminal is too small. 

From your first comment:

>     * On small screens we should fall back to the current behavior, which is more compact (always takes 1 row). When to fall back exactly? A few random ideas (feel free to suggest more):

osa1 replied:
> I just thought you wanted it to fall back to scrolling when the terminal is too small.

I still do, but I'm not sure if this is the best way to do it.

I guess it's fine for now. I just tried again and it seems to work as advertised.

trevarj commented on libtiny_tui/src/messaging.rs:
```diff
@@ -113,7 +113,17 @@ impl MessagingUI {
     }
 
-    pub(crate) fn draw(&self, tb: &mut Termbox, colors: &Colors, pos_x: i32, pos_y: i32) {
+    pub(crate) fn draw(&mut self, tb: &mut Termbox, colors: &Colors, pos_x: i32, pos_y: i32) {
+        let lines = self.input_field.calculate_lines();
+
+        // make space for textfield expansion
+        if self.height - lines >= 5 {
+            self.msg_area.resize(self.width, self.height - lines);
+        } else {
+            let widget_width = self.calculate_input_field_width(self.width);
+            self.input_field.resize(widget_width, 1);
+        }
+
```
Yes, I agree...It was just the only thing I could think of to make the elements dynamically resize 😞

osa1 commented:
@trevarj in my "extensive" testing of this PR I found a bug in rendering of ExitDialogue (unrelated to this patch), which I just fixed on master. Unfortunately it caused a merge conflict in this PR. Would you be able to fix the conflict? If not let me know and I'm going to fix it myself.

The change is simple: I removed `draw_input_field` method of `MessagingUI` and very slightly refactored (though indentation changed quite a bit) the `draw` method. We now don't show the nick when rendering `ExitDialogue` so it now takes full width.

Sorry again for the trouble..

trevarj commented:
> 
> 
> @trevarj in my "extensive" testing of this PR I found a bug in rendering of ExitDialogue (unrelated to this patch), which I just fixed on master. Unfortunately it caused a merge conflict in this PR. Would you be able to fix the conflict? If not let me know and I'm going to fix it myself.
> 
> The change is simple: I removed `draw_input_field` method of `MessagingUI` and very slightly refactored (though indentation changed quite a bit) the `draw` method. We now don't show the nick when rendering `ExitDialogue` so it now takes full width.
> 
> Sorry again for the trouble..

No problem, I will take care of it.

osa1 commented:
Thanks!

osa1 commented:
Unfortunately because you did a merge rather than rebase I can't squash these commits now. Anyways I'll be rebasing in my staging branch.

trevarj commented:
Sorry! 😞 Do you want me to revert?

osa1 commented:
No need! Already rebased in `textfield-overflow` branch.

osa1 commented:
Sigh.. I keep finding bugs:

![Screenshot_2020-04-12_15-33-14](https://user-images.githubusercontent.com/448274/79068900-f397a700-7cd2-11ea-9510-68f7a99334bb.png)

I'll give instructions for reproducing in a bit. Let's continue development and bug fixing in a branch that I can commit to. @trevarj would you like to give me commit access to your repo so that I can update your branch?

trevarj commented:
Should I close this PR since the changes are already in your repo?

osa1 commented:
We can close this and create a new PR from my branch to continue discussion. Or just continue here. I don't have a preference.

osa1 commented:
Let's continue from this branch as both of us have push access here. I'll have to give you push access to the main repo (to all branches, as far as I can see) if we want to collaborate on this patch using a branch on my repo.

trevarj commented:
> 
> 
> Let's continue from this branch as both of us have push access here. I'll have to give you push access to the main repo (to all branches, as far as I can see) if we want to collaborate on this patch using a branch on my repo.

Ok, sounds good.

osa1 commented:
Here's the backtrace when the bug I mentioned before happens:

<details>

<summary>Backtrace</summary>

```
thread 'main' panicked at 'assertion failed: y >= 0', termbox/src/lib.rs:342:9
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1069
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1504
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:218
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:511
  11: std::panicking::begin_panic
             at /home/omer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:438
  12: termbox_simple::Termbox::change_cell
             at termbox/src/lib.rs:342
  13: libtiny_tui::termbox::print_chars
             at libtiny_tui/src/termbox.rs:11
  14: libtiny_tui::text_field::draw_line_scroll
             at libtiny_tui/src/text_field.rs:645
  15: libtiny_tui::text_field::draw_line
             at libtiny_tui/src/text_field.rs:673
  16: libtiny_tui::text_field::TextField::draw
             at libtiny_tui/src/text_field.rs:89
  17: libtiny_tui::messaging::MessagingUI::draw
             at libtiny_tui/src/messaging.rs:148
  18: libtiny_tui::tui::TUI::draw
             at libtiny_tui/src/tui.rs:779
  19: libtiny_tui::tui::TUI::resize_
             at libtiny_tui/src/tui.rs:669
  20: libtiny_tui::tui::TUI::resize
             at libtiny_tui/src/tui.rs:597
  21: libtiny_tui::sigwinch_handler::{{closure}}
             at libtiny_tui/src/lib.rs:79
  22: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/omer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/future/mod.rs:66
  23: tokio::task::core::Core<T>::poll
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/task/core.rs:128
  24: tokio::task::harness::Harness<T,S>::poll::{{closure}}::{{closure}}
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/task/harness.rs:119
  25: core::ops::function::FnOnce::call_once
             at /home/omer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/ops/function.rs:232
  26: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /home/omer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panic.rs:318
  27: std::panicking::try::do_call
             at /home/omer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:331
  28: __rust_try
  29: std::panicking::try
             at /home/omer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:274
  30: std::panic::catch_unwind
             at /home/omer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panic.rs:394
  31: tokio::task::harness::Harness<T,S>::poll::{{closure}}
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/task/harness.rs:100
  32: tokio::loom::std::causal_cell::CausalCell<T>::with_mut
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/loom/std/causal_cell.rs:41
  33: tokio::task::harness::Harness<T,S>::poll
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/task/harness.rs:99
  34: tokio::task::raw::poll
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/task/raw.rs:162
  35: tokio::task::raw::RawTask::poll
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/task/raw.rs:113
  36: tokio::task::Task<S>::run
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/task/mod.rs:381
  37: tokio::task::local::Scheduler::tick
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/task/local.rs:513
  38: <tokio::task::local::LocalFuture<F> as core::future::future::Future>::poll::{{closure}}
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/task/local.rs:407
  39: tokio::task::local::Scheduler::with::{{closure}}
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/task/local.rs:478
  40: std::thread::local::LocalKey<T>::try_with
             at /home/omer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/thread/local.rs:263
  41: std::thread::local::LocalKey<T>::with
             at /home/omer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/thread/local.rs:239
  42: tokio::task::local::Scheduler::with
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/task/local.rs:474
  43: <tokio::task::local::LocalFuture<F> as core::future::future::Future>::poll
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/task/local.rs:402
  44: core::future::poll_with_context
             at /home/omer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/future/mod.rs:84
  45: tokio::task::local::LocalSet::run_until::{{closure}}
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/task/local.rs:358
  46: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/omer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/future/mod.rs:66
  47: tokio::runtime::basic_scheduler::BasicScheduler<P>::block_on
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/runtime/basic_scheduler.rs:138
  48: tokio::runtime::Runtime::block_on::{{closure}}
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/runtime/mod.rs:413
  49: tokio::runtime::context::enter
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/runtime/context.rs:72
  50: tokio::runtime::handle::Handle::enter
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/runtime/handle.rs:34
  51: tokio::runtime::Runtime::block_on
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/runtime/mod.rs:410
  52: tokio::task::local::LocalSet::block_on
             at /home/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.13/src/task/local.rs:321
  53: chat::main
             at libtiny_tui/examples/chat.rs:30
  54: std::rt::lang_start::{{closure}}
             at /home/omer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67
  55: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:52
  56: std::panicking::try::do_call
             at src/libstd/panicking.rs:331
  57: std::panicking::try
             at src/libstd/panicking.rs:274
  58: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  59: std::rt::lang_start_internal
             at src/libstd/rt.rs:51
  60: std::rt::lang_start
             at /home/omer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67
  61: main
  62: __libc_start_main
  63: _start
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

</details>

The relevant bits:

```
  12: termbox_simple::Termbox::change_cell
             at termbox/src/lib.rs:342
  13: libtiny_tui::termbox::print_chars
             at libtiny_tui/src/termbox.rs:11
  14: libtiny_tui::text_field::draw_line_scroll
             at libtiny_tui/src/text_field.rs:645
  15: libtiny_tui::text_field::draw_line
             at libtiny_tui/src/text_field.rs:673
  16: libtiny_tui::text_field::TextField::draw
             at libtiny_tui/src/text_field.rs:89
  17: libtiny_tui::messaging::MessagingUI::draw
             at libtiny_tui/src/messaging.rs:148
```

So it seems like this `print_chars` call:

```
fn draw_line_scroll(
    tb: &mut Termbox,
    colors: &Colors,
    line: &[char],
    pos_x: i32,
    pos_y: i32,
    width: i32,
    cursor: i32,
    scroll: i32,
) {
    let slice: &[char] = &line[scroll as usize..min(line.len(), (scroll + width) as usize)];
    termbox::print_chars(tb, pos_x, pos_y, colors.user_msg, slice.iter().cloned());
    ...
}
```

is passing negative indices in some cases.

Here's how I'm getting this panic:

- Make sure your branch is up-to-date (I did some force pushes so make sure to reset your branch)
- In `libtiny_tui` run `RUST_BACKTRACE=1 cargo run --example chat 2>stderr`
- Switch to a tab with messages, then type something long.
- Make your window tiny and just keep resizing (but keep the window small) until things get weird.
- You should now see a `stderr` file with the backtrace shown above.

@trevarj could you take a look?

osa1 commented:
I think we could catch these kinds of errors earlier if we had better types for fields like `width`, `max_lines` etc. It doesn't make sense for these numbers to be negative, so I think unsized integer types would model them better. With overflow and underflow checking (enabled by default in debug builds) it should make debugging easier. But perhaps that's a task for another PR..

osa1 commented:
I'm also a bit surprised that the panic above *does not* cause tiny to abort. No idea why and how it keeps running..

trevarj commented:
Changing all the `i32` to `u32` seems like a large undertaking right now 😓 , but I will look into changing `width` to `u32` and/or adding safety around the calculations in those functions.

trevarj commented:
I get weird results on certain terminal emulators. I tried a few (gnome terminal, konsole, alacritty, mintty for WSL), but was able to reproduce on mintty for Windows every time.

Can you cross-check this fix on your konsole? Using konsole gave no issues with gnome as my DE.

change line 353 of termbox/src/lib.rs to :
`tty.write_all(&self.output_buffer).unwrap_or_default();`

osa1 commented:
Hmm, is this line really related to the bug I'm observing? If you look at the backtrace, the panic is happening because we pass negative numbers termbox's `change_cell`.

trevarj commented:
Well, I originally did some protection around `change_cell` starting in `draw_line_scroll` by safely casting to `u32` or defaulting to 0 if negative. Then I was able to find another stacktrace at the line above. I decided to remove the casts to `u32` because it's very messy and ultimately needs to be done throughout the entire UI, and it still seems to solve the issue (on my end at least).

osa1 commented:
Interesting. Let me try to find a reliable reproducer a little bit more.

trevarj commented:
Can you test with this fix? https://github.com/trevarj/tiny/tree/small-window-bug

osa1 commented:
That should fix it as you no longer pass negative numbers to termbox, but I think it's a better idea to fix the disease (computing a negative number for location of a character in TUI) rather than the symptom (assertion error in termbox). I'm at work currently but I'll try to see why (under what conditions) we pass negative numbers in the current code later.

trevarj commented:
You're right. 
Now, I'm looking at line 120 in messaging.rs :
`let input_field_y = pos_y + self.height - lines;`
The minimum for `lines` is always 1, so when `pos_y ` and `self.height` are 0 `input_field_y ` will be -1.

osa1 commented:
OK so I've been using this feature today and there are two things that I'm not quite happy about. Good news is that I think it's possible to make this both simpler (in code) and better (in UX).

There are two problems:

- Splitting the line at the last character (instead of at word boundary) looks bad. Reading something like
  ```
  osa1: blah blah b
        lah blah
  ```
  is harder than
  ```
  osa1: blah blah
        blah blah
  ```

- Indenting lines looks cool, but when we send the message it's shown without indentation in consecutive lines. So for example, if I'm typing this:
  ```
  osa1: blah blah blah
        blah blah blah
  ```
  when I send this the message it looks like this in the message area:
  ```
  osa1: blah blah blah
  blah blah blah
  ```
  (Note that the second line is not indented. We have a feature request for aligning all messages: #169)

I think it's possible to fix both and make the code simpler:

- First of all decide whether to wrap or scroll. Let's keep the logic simple (easier to code, easier to explain, easier to test). Idea:
    - We will always show the current nick in wrapping mode (already the case currently)
    - Consider how many lines the text field will take. This is simply by taking length of the nick + text field and splitting the lines at word boundaries, similar to how we do it when adding new lines to messaging area. We don't want to compute this on every `draw()` so we'll probably have to cache some state here.
    - If the text field will take more than half of the screen height then fall back to scrolling.

   Crucially this idea doesn't care about nick length in wrap mode, it just treats the input field as a whole, including the nick part.

- Rendering in scroll mode is the same as before.

- Rendering in wrap mode is done the same way we render a line in text field. Perhaps we could even re-use the method `Line::draw()` (perhaps with some changes, maybe make it a function etc.).

Any thoughts @trevarj ?

I'm happy to implement this (based on your code in this PR) or at least help. Just let me know.

trevarj commented:
@osa1 
I think fixing these problems will definitely improve UX. I'm ready to dig in and make the changes, but maybe you can guide me a little with the low-level design.  I don't want to do a bunch of bad work and have it be opposite of what you have in mind.

If we don't want to do the refactoring, then I can just get started on splitting lines on whitespace.

trevarj commented:
Would you be opposed to making `MsgArea` like this? : 

```
03:04        osa1 : sdfsdfsdfsd
03:05        osa1 : sdfsdfsdfsdfsdf              
                    sdfsdsdf
03:05     trevarj : sdfsdfs
```

trevarj commented:
I tried to implement `textwrap` https://docs.rs/textwrap/0.11.0/textwrap/index.html
on the draw_line() so that the words wrap, but I found it really difficult to draw the cursor in the correct place, and also pre-determine the number of lines the TextField will have in `MessageUI::draw()` 😖

osa1 commented:
> but I found it really difficult to draw the cursor in the correct place

Hmm, I'm afraid I don't see why this is difficult. Here's what I have in mind: when rendering the text field we already now the cursor position. Render it char-by-char while counting current char index (use `enumerate()`). If the current char index is the same as cursor index then render cursor. Otherwise render as usual.

I'll take a look at the code and see what the problem is.

> also pre-determine the number of lines the TextField will have in MessageUI::draw()

I think this should be calculated when we call `TextField::resize()`. This is what we do for `Line`s in the messaging area, and the logic should be more-or-less the same, I think.

trevarj commented:
I've put up a PR over on my fork ([here](https://github.com/trevarj/tiny/pull/1) with a working solution to the input field wrapping on words. It's working nicely as far as I can tell...even works with autocompletion.
I'm currently looking at the indentation for messages in the `MsgArea` to see if there's an easy solution. I need to study `MsgArea` and `Line` more to see if I can use the code I wrote for `TextField`.

trevarj commented:
@osa1 have you gotten a chance to play around with this latest build? I'm interested in your thoughts after using it for a bit.

osa1 commented:
I just installed the latest version. Let me use it for a bit, I'll update.

osa1 commented:
I've been using this for the last couple of days. Here's a screenshot of this patch in action: (text field is empty as that's the same as before, only the changes in the messaging area are visible in this SS)

![Screenshot_2020-05-09_17-16-28](https://user-images.githubusercontent.com/448274/81476215-5704ee00-9219-11ea-9f9c-328ea59d3f8c.png)

The main issue is that this does opposite of what I suggested above (sorry if my wording was not clear enough), namely, it aligns everything, rather than not aligning user input. I think it doesn't look good even though it's consistent (both input field and messaging area are formatted the same way), but more importantly tiny is supposed to compact, and for that I'd much rather skip the alignment and start consecutive lines at column 0.

Only bug I could find is also visible in this SS, if you look at the line with the long link you'll see it's in a new line for no reason.

I'm busy with work and other stuff these days so I couldn't review this line by line, but here's a suggestion (feel free to ignore if you think it wouldn't work or would be too much work): implement something like `Line` that support these operations:

- `O(n)`-time rendering: it should be possible to simply scan it left-to-right and add one char at a time (or move on to the next line). For this you'll need to maintain indices of whitespace or maybe split the input into segments, as we do in `Line`.
- Should support adding/removing chars (for user input). It's fine if this does vector insertion in the middle or something like that, but adding a new character to the end (the most common operation) should be `O(1)` (this should be easily possible, I think).
- Should support pasting. It's fine if we create the entire line (with segments/whitespace indices etc.) from scratch when this happens.
- Unless I'm missing something, this type should not need a `resize()` method, similar to how we don't need `resize()` for `Line`.
- It should support starting at the given column when rendering the first line. This will be used to render nick before the input field. E.g. if the nick string is "osa1: " we'll do something like `input_field.draw(tui, x_pos, y_pos, 5)` where `5` indicates that in the first line it'll skip the first 5 columns (so the first character will be at `x_pos + 5`). However in the next line we should be rendering starting from `pos_x`, not `pox_x + 5`!

This code should live in its own module.

Once we have that it'll be trivial to replace the current `TextField` with this one: we'll simply replace the field in `MessagingUI`, and all the mess and important stuff to carefully implement and review will be in a single module.

Btw, I just realized that it's not possible to implement this feature without calling `resize` in some cases on user input (something that I wanted to avoid so far). The reason is this: suppose the text field is at its full width and I add one more character. Now the text field will grow one more line, and for that messaging area needs to shrink by one line, so we'll have to resize it. I think this is fine as there's no way around this.

One edge case here is what happens if I keep adding more characters and messaging area becomes too small. I think we should handle this by falling back to the scroll mode.

The actual resizing logic (when terminal size changes) should be the one I describe above.

osa1 commented:
I forgot to mention, thanks for your persistence and sorry that I've not been more responsive. One problem is I do code reviews at work, so that's not something I've been super keen on doing in my free time. But I'm actually keen on merging this! Hopefully we'll get there.

trevarj commented:
> One problem is I do code reviews at work, so that's not something I've been super keen on doing in my free time.
It's not a problem. Thanks for guiding me through this. I will admit that I am a bit weary of doing a full re-factor because I'm worried I don't have a good idea of what you have in mind. It is a good Rust learning experience for me though, so I will try to attempt something.

trevarj commented:
Hi @osa1 ,
I'm working on this refactor over here: https://github.com/trevarj/tiny/tree/text-wrap-refactor
Right now it has the basic functionality done. Autocompletion is not working yet, and I haven't thoroughly tested everything. 

Do you think you could take a look at it when you have time? I don't want to do too much and go off track of what you are envisioning.

High level overview:

-  Migrated `text_field.rs` into its own module - `input_area`
   - `mod.rs` - contains `struct TextField` and all related functions
   - `input_line.rs` - contains a new struct `InputLine` which will replace `TextField`'s buffer (similar to `Line` in `msg_area`
      - To track whitespaces/splits I decided to use a BTreeSet to make it slightly easier for inserting/removing, something that `msg_area::Line` doesn't need to do. I'd like to know your thoughts on the performance of that vs a `Vec`.
- Moved nickname to inside `TextField` to have it draw the nick instead of in `MessagingUI`

Thank you.

osa1 commented:
Hey @trevarj,

Just took a look -- the code looks great, thanks! It's easy to review the changes in `messaging.rs` as important bits are all in their own file now.

I found some bugs while testing but you said you haven't thoroughly tested everything so not reporting those now.

> Moved nickname to inside TextField to have it draw the nick instead of in MessagingUI

This makes sense to me :+1:

Regarding BTreeSet vs. Vec: I think nothing's going to beat Vec iteration performance, but looking at the code, it doesn't seem too hard to replace the BTreeSet with Vec if needed later. So use whatever's most convenient for you, we can replace it later if needed, using some benchmarks to guide the changes.

I think we could cache `calculate_lines` output. Invalidate it on resize() and when input changes, in all other cases it could return the previous calculated number.

It'd great if you could rebase your branch on `master` (not a merge!), that makes reviewing so much easier and it'll also make merging easier when the time comes.

Overall looks great! Thanks again!

trevarj commented:
New conversation here: https://github.com/osa1/tiny/pull/200
